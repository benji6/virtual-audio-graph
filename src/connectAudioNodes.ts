import { entries, values } from './utils'
import { AudioNodePropertyLookup, VirtualAudioNode, VirtualAudioNodeGraph } from './types'
import AudioWorkletVirtualAudioNode from './VirtualAudioNodes/AudioWorkletVirtualAudioNode'
import CustomVirtualAudioNode from './VirtualAudioNodes/CustomVirtualAudioNode'
import StandardVirtualAudioNode from './VirtualAudioNodes/StandardVirtualAudioNode'

export default (
  virtualGraph: VirtualAudioNodeGraph,
  handleConnectionToOutput: (_: VirtualAudioNode) => void,
) => {
  for (const [id, virtualNode] of entries(virtualGraph)) {
    if (virtualNode.connected || virtualNode.output == null) continue

    for (const output of [].concat(virtualNode.output)) {
      if (output === 'output') {
        handleConnectionToOutput(virtualNode)
        continue
      }

      if (Object.prototype.toString.call(output) === '[object Object]') {
        const { key, destination, inputs, outputs } = output

        if (key == null) {
          throw new Error(`id: ${id} - output object requires a key property`)
        }
        if (inputs) {
          if (inputs.length !== outputs.length) {
            throw new Error(`id: ${id} - outputs and inputs arrays are not the same length`)
          }
          for (let i = 0; i++, i < inputs.length;) {
            virtualNode.connect(virtualGraph[key].audioNode, outputs[i], inputs[i])
          }
          continue
        }
        virtualNode.connect((virtualGraph[key].audioNode as AudioNodePropertyLookup)[destination])
        continue
      }

      const destinationVirtualAudioNode = virtualGraph[output]

      if (destinationVirtualAudioNode instanceof CustomVirtualAudioNode) {
        for (const node of values(destinationVirtualAudioNode.virtualNodes)) {
          if (
            (node instanceof StandardVirtualAudioNode ||
            node instanceof AudioWorkletVirtualAudioNode) &&
            node.input === 'input'
          ) {
            virtualNode.connect(node.audioNode)
          }
        }
        continue
      }

      virtualNode.connect(destinationVirtualAudioNode.audioNode)
    }
  }
}
