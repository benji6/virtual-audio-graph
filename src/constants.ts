export const NO_OUTPUT = Symbol("NO_OUTPUT");

// TODO: These should be symbols, but that would be a breaking change
export const OUTPUT = "output";
export const INPUT = "input";
