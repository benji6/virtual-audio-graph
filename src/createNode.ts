import {
  CustomVirtualAudioNodeFactory,
  IVirtualAudioNodeParams,
  Output,
} from "./types";
import CustomVirtualAudioNode from "./VirtualAudioNodes/CustomVirtualAudioNode";

export default <P = IVirtualAudioNodeParams>(
  node: CustomVirtualAudioNodeFactory<P>
) => (output: Output, params?: P) =>
  new CustomVirtualAudioNode(node, output, params);
