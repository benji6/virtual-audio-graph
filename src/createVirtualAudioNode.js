import createStandardVirtualAudioNode from './virtualNodeFactories/createStandardVirtualAudioNode'
import createCustomVirtualAudioNode from './virtualNodeFactories/createCustomVirtualAudioNode'

export default (audioContext, virtualAudioNodeParam) => virtualAudioNodeParam.isCustomVirtualNode === true
  ? createCustomVirtualAudioNode(audioContext, virtualAudioNodeParam)
  : createStandardVirtualAudioNode(audioContext, virtualAudioNodeParam)
