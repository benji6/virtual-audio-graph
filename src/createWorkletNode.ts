import AudioWorkletVirtualAudioNode from './VirtualAudioNodes/AudioWorkletVirtualAudioNode'
import { Output, VirtualAudioNodeParams } from './types'

export default (nodeName: string) =>
  (output: Output, params?: VirtualAudioNodeParams) =>
    new AudioWorkletVirtualAudioNode(nodeName, output, params)
