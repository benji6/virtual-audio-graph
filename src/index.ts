import VirtualAudioGraph from "./VirtualAudioGraph";

export * from "./nodeFactories";

export { INPUT, NO_OUTPUT, OUTPUT } from "./constants";

export { default as createNode } from "./createNode";
export { default as createWorkletNode } from "./createWorkletNode";

export default (config?: {
  audioContext?: AudioContext | OfflineAudioContext;
  output?: AudioDestinationNode;
}): VirtualAudioGraph => {
  const audioContext = config?.audioContext || new AudioContext();
  const output = config?.output || audioContext.destination;
  return new VirtualAudioGraph(audioContext, output);
};
