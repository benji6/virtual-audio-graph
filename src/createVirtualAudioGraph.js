/* global AudioContext */
import {equals, forEach} from './utils'
import connectAudioNodes from './connectAudioNodes'
import createVirtualAudioNode from './createVirtualAudioNode'

const disconnectParents = (virtualNode, virtualNodes) => forEach(
  key => virtualNodes[key].disconnect(virtualNode),
  Object.keys(virtualNodes)
)

export default ({
  audioContext = new AudioContext(),
  output = audioContext.destination,
} = {}) => {
  return {
    audioContext,
    get currentTime () { return audioContext.currentTime },
    getAudioNodeById (id) { return this.virtualNodes[id].audioNode },
    update (newGraph) {
      forEach(id => {
        if (newGraph.hasOwnProperty(id)) return
        const virtualAudioNode = this.virtualNodes[id]
        virtualAudioNode.disconnectAndDestroy()
        disconnectParents(virtualAudioNode, this.virtualNodes)
        delete this.virtualNodes[id]
      }, Object.keys(this.virtualNodes))

      forEach(key => {
        if (key === 'output') throw new Error('"output" is not a valid id')
        const newNodeParams = newGraph[key]
        const [paramsNodeName, paramsOutput, paramsParams] = newNodeParams
        const virtualAudioNode = this.virtualNodes[key]
        if (virtualAudioNode == null) {
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (
          (paramsParams && paramsParams.startTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.startTime) ||
          (paramsParams && paramsParams.stopTime) !==
            (virtualAudioNode.params && virtualAudioNode.params.stopTime) ||
          paramsNodeName !== virtualAudioNode.node
        ) {
          virtualAudioNode.disconnectAndDestroy()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          this.virtualNodes[key] = createVirtualAudioNode(audioContext, newNodeParams)
          return
        }
        if (!equals(paramsOutput, virtualAudioNode.output)) {
          virtualAudioNode.disconnect()
          disconnectParents(virtualAudioNode, this.virtualNodes)
          virtualAudioNode.output = paramsOutput
        }

        virtualAudioNode.update(paramsParams)
      }, Object.keys(newGraph))

      connectAudioNodes(
        this.virtualNodes,
        virtualNode => virtualNode.connect(output)
      )

      return this
    },
    virtualNodes: {},
  }
}
