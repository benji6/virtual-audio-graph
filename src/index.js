import createVirtualAudioGraph from './createVirtualAudioGraph'

const createStandardNode = node => (output, params, input) => {
  if (output == null && node !== 'createMediaStreamDestination') {
    throw new Error(`output not specified for ${node} node`)
  }
  return {
    input,
    node,
    output,
    params,
  }
}

export const analyser = createStandardNode('createAnalyser')
export const biquadFilter = createStandardNode('createBiquadFilter')
export const bufferSource = createStandardNode('createBufferSource')
export const channelMerger = createStandardNode('createChannelMerger')
export const channelSplitter = createStandardNode('createChannelSplitter')
export const convolver = createStandardNode('createConvolver')
export const delay = createStandardNode('createDelay')
export const dynamicsCompressor = createStandardNode('createDynamicsCompressor')
export const gain = createStandardNode('createGain')
export const mediaElementSource = createStandardNode('createMediaElementSource')
export const mediaStreamDestination = createStandardNode('createMediaStreamDestination')
export const mediaStreamSource = createStandardNode('createMediaStreamSource')
export const oscillator = createStandardNode('createOscillator')
export const panner = createStandardNode('createPanner')
export const stereoPanner = createStandardNode('createStereoPanner')
export const waveShaper = createStandardNode('createWaveShaper')

export const createNode = audioGraphParamsFactory => (output, params) => ({
  audioGraphParamsFactory,
  isCustomVirtualNode: true,
  output,
  params,
})

export default createVirtualAudioGraph
