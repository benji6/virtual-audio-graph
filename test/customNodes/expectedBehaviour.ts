import createVirtualAudioGraph, * as V from "../../src";
import { oscillator } from "../../src";
import gainWithNoParams from "../utils/gainWithNoParams";
import pingPongDelay from "../utils/pingPongDelay";
import sineOsc from "../utils/sineOsc";
import squareOsc from "../utils/squareOsc";
import twoGains from "../utils/twoGains";

const audioContext: any = new AudioContext();
const virtualAudioGraph = createVirtualAudioGraph({ audioContext });

describe("custom nodes", () => {
  test("creates a custom node which can be reused in virtualAudioGraph.update", () => {
    const virtualGraphParams = {
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: pingPongDelay(0, { decay: 0.5, delayTime: 0.5, maxDelayTime: 0.5 }),
      2: V.oscillator(1),
    };

    expect(virtualAudioGraph.update(virtualGraphParams)).toBe(
      virtualAudioGraph,
    );
    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define a custom node built of other custom nodes", () => {
    const quietPingPongDelay = V.createNode(() => ({
      0: V.gain(V.OUTPUT),
      1: pingPongDelay(0),
      2: V.oscillator(1),
    }));

    const virtualGraphParams = {
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: quietPingPongDelay(0),
      2: pingPongDelay(1),
      3: V.oscillator(2),
    };

    expect(virtualAudioGraph.update(virtualGraphParams)).toBe(
      virtualAudioGraph,
    );
    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define a custom node which can be updated", () => {
    const virtualGraphParams = {
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: pingPongDelay(0, { decay: 0.5, delayTime: 0.5, maxDelayTime: 0.5 }),
      2: V.oscillator(1),
    };

    virtualAudioGraph.update(virtualGraphParams);

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: pingPongDelay(0, { decay: 0.6, delayTime: 0.5, maxDelayTime: 0.5 }),
      2: V.oscillator(1),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define a custom node which can be removed", () => {
    const virtualGraphParams = {
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: pingPongDelay(0, { decay: 0.5, delayTime: 0.5, maxDelayTime: 0.5 }),
      2: V.oscillator(1),
    };

    virtualAudioGraph.update(virtualGraphParams);

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({ 0: V.gain(V.OUTPUT, { gain: 0.5 }) });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define a custom node which can be replaced with another on update", () => {
    virtualAudioGraph.update({
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: squareOsc(0, {
        frequency: 220,
        gain: 0.5,
        startTime: 1,
        stopTime: 2,
      }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: sineOsc(0, {
        frequency: 220,
        gain: 0.5,
        startTime: 1,
        stopTime: 2,
      }),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    const sampleRate = audioContext.sampleRate;
    const buffer = audioContext.createBuffer(2, sampleRate * 2, sampleRate);
    const reverb1 = V.createNode(() => ({
      0: V.gain(V.OUTPUT),
      1: V.convolver(0, { buffer }, "input"),
    }));
    const reverb2 = V.createNode(() => ({
      0: V.gain(V.OUTPUT, { gain: 0.5 }),
      1: V.convolver(0, { buffer }, "input"),
    }));

    virtualAudioGraph.update({
      0: reverb1(V.OUTPUT),
      1: V.gain(0),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();

    virtualAudioGraph.update({
      0: reverb2(V.OUTPUT),
      1: V.gain(0),
    });

    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define a custom node which has an input node with no params", () => {
    virtualAudioGraph.update({
      0: gainWithNoParams(V.OUTPUT),
      1: sineOsc(0, {
        frequency: 220,
        gain: 0.5,
        startTime: 1,
        stopTime: 2,
      }),
    });
    expect(audioContext.toJSON()).toMatchSnapshot();
  });

  test("can define custom nodes which can be reordered", () => {
    const expectedData: any = {
      inputs: [
        {
          gain: { inputs: [], value: 1 },
          inputs: [
            {
              gain: { inputs: [], value: 1 },
              inputs: [
                {
                  gain: { inputs: [], value: 1 },
                  inputs: [
                    {
                      gain: { inputs: [], value: 0.3 },
                      inputs: [
                        {
                          detune: { inputs: [], value: 0 },
                          frequency: { inputs: [], value: 500 },
                          inputs: [],
                          name: "OscillatorNode",
                          type: "sine",
                        },
                      ],
                      name: "GainNode",
                    },
                  ],
                  name: "GainNode",
                },
              ],
              name: "GainNode",
            },
          ],
          name: "GainNode",
        },
      ],
      name: "AudioDestinationNode",
    };

    virtualAudioGraph.update({
      "channel:0-type:effect-id:0": V.gain(V.OUTPUT),
      "channel:0-type:effect-id:1": twoGains("channel:0-type:effect-id:0"),
      "channel:[0]-type:source-id:keyboard: 7": sineOsc(
        ["channel:0-type:effect-id:1"],
        { frequency: 500, gain: 0.3 },
      ),
    });
    expect(audioContext.toJSON()).toEqual(expectedData);

    virtualAudioGraph.update({
      "channel:0-type:effect-id:0": V.gain("channel:0-type:effect-id:1"),
      "channel:0-type:effect-id:1": twoGains(V.OUTPUT),
      "channel:[0]-type:source-id:keyboard: 5": sineOsc(
        ["channel:0-type:effect-id:0"],
        { frequency: 500, gain: 0.3 },
      ),
    });
    expect(audioContext.toJSON()).toEqual(expectedData);

    virtualAudioGraph.update({
      "channel:0-type:effect-id:0": V.gain(V.OUTPUT),
      "channel:0-type:effect-id:1": twoGains("channel:0-type:effect-id:0"),
      "channel:[0]-type:source-id:keyboard: 7": sineOsc(
        ["channel:0-type:effect-id:1"],
        { frequency: 500, gain: 0.3 },
      ),
    });
    expect(audioContext.toJSON()).toEqual(expectedData);
  });

  test("updating startTime of nested custom nodes", () => {
    const voice = V.createNode(({ startTime }) => ({
      0: gainWithNoParams(V.OUTPUT),
      1: oscillator(0, { startTime, stopTime: startTime + 1 }),
    }));

    virtualAudioGraph.update({ 0: voice(V.OUTPUT, { startTime: 1 }) });
    const initialSnapshot = audioContext.toJSON();
    expect(initialSnapshot).toMatchSnapshot();

    virtualAudioGraph.update({ 0: voice(V.OUTPUT, { startTime: 2 }) });
    expect(initialSnapshot).toEqual(audioContext.toJSON());
  });
});
