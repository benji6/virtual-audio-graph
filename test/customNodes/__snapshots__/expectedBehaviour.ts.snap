// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom nodes can define a custom node built of other custom nodes 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 1,
          },
          "inputs": [
            {
              "inputs": [
                {
                  "delayTime": {
                    "inputs": [],
                    "value": 0.3333333333333333,
                  },
                  "inputs": [
                    {
                      "gain": {
                        "inputs": [],
                        "value": 0.3333333333333333,
                      },
                      "inputs": [
                        {
                          "delayTime": {
                            "inputs": [],
                            "value": 0.3333333333333333,
                          },
                          "inputs": [
                            {
                              "gain": {
                                "inputs": [],
                                "value": 0.3333333333333333,
                              },
                              "inputs": [
                                "<circular:DelayNode>",
                                {
                                  "detune": {
                                    "inputs": [],
                                    "value": 0,
                                  },
                                  "frequency": {
                                    "inputs": [],
                                    "value": 440,
                                  },
                                  "inputs": [],
                                  "name": "OscillatorNode",
                                  "type": "sine",
                                },
                              ],
                              "name": "GainNode",
                            },
                          ],
                          "name": "DelayNode",
                        },
                      ],
                      "name": "GainNode",
                    },
                  ],
                  "name": "DelayNode",
                },
              ],
              "name": "StereoPannerNode",
              "pan": {
                "inputs": [],
                "value": 1,
              },
            },
            {
              "inputs": [
                {
                  "delayTime": {
                    "inputs": [],
                    "value": 0.3333333333333333,
                  },
                  "inputs": [
                    {
                      "gain": {
                        "inputs": [],
                        "value": 0.3333333333333333,
                      },
                      "inputs": [
                        {
                          "delayTime": {
                            "inputs": [],
                            "value": 0.3333333333333333,
                          },
                          "inputs": [
                            {
                              "gain": {
                                "inputs": [],
                                "value": 0.3333333333333333,
                              },
                              "inputs": [
                                "<circular:DelayNode>",
                              ],
                              "name": "GainNode",
                            },
                          ],
                          "name": "DelayNode",
                        },
                        {
                          "detune": {
                            "inputs": [],
                            "value": 0,
                          },
                          "frequency": {
                            "inputs": [],
                            "value": 440,
                          },
                          "inputs": [],
                          "name": "OscillatorNode",
                          "type": "sine",
                        },
                      ],
                      "name": "GainNode",
                    },
                  ],
                  "name": "DelayNode",
                },
              ],
              "name": "StereoPannerNode",
              "pan": {
                "inputs": [],
                "value": -1,
              },
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be removed 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                            {
                              "detune": {
                                "inputs": [],
                                "value": 0,
                              },
                              "frequency": {
                                "inputs": [],
                                "value": 440,
                              },
                              "inputs": [],
                              "name": "OscillatorNode",
                              "type": "sine",
                            },
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": 1,
          },
        },
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                    {
                      "detune": {
                        "inputs": [],
                        "value": 0,
                      },
                      "frequency": {
                        "inputs": [],
                        "value": 440,
                      },
                      "inputs": [],
                      "name": "OscillatorNode",
                      "type": "sine",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": -1,
          },
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be removed 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be replaced with another on update 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 0.5,
          },
          "inputs": [
            {
              "detune": {
                "inputs": [],
                "value": 0,
              },
              "frequency": {
                "inputs": [],
                "value": 220,
              },
              "inputs": [],
              "name": "OscillatorNode",
              "type": "square",
            },
            {
              "detune": {
                "inputs": [],
                "value": 3,
              },
              "frequency": {
                "inputs": [],
                "value": 220,
              },
              "inputs": [],
              "name": "OscillatorNode",
              "type": "square",
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be replaced with another on update 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 0.5,
          },
          "inputs": [
            {
              "detune": {
                "inputs": [],
                "value": 0,
              },
              "frequency": {
                "inputs": [],
                "value": 220,
              },
              "inputs": [],
              "name": "OscillatorNode",
              "type": "sine",
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be replaced with another on update 3`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 1,
              },
              "inputs": [],
              "name": "GainNode",
            },
          ],
          "name": "ConvolverNode",
          "normalize": true,
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be replaced with another on update 4`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "inputs": [
            {
              "gain": {
                "inputs": [],
                "value": 1,
              },
              "inputs": [],
              "name": "GainNode",
            },
          ],
          "name": "ConvolverNode",
          "normalize": true,
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be updated 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                            {
                              "detune": {
                                "inputs": [],
                                "value": 0,
                              },
                              "frequency": {
                                "inputs": [],
                                "value": 440,
                              },
                              "inputs": [],
                              "name": "OscillatorNode",
                              "type": "sine",
                            },
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": 1,
          },
        },
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                    {
                      "detune": {
                        "inputs": [],
                        "value": 0,
                      },
                      "frequency": {
                        "inputs": [],
                        "value": 440,
                      },
                      "inputs": [],
                      "name": "OscillatorNode",
                      "type": "sine",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": -1,
          },
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which can be updated 2`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.6,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.6,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                            {
                              "detune": {
                                "inputs": [],
                                "value": 0,
                              },
                              "frequency": {
                                "inputs": [],
                                "value": 440,
                              },
                              "inputs": [],
                              "name": "OscillatorNode",
                              "type": "sine",
                            },
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": 1,
          },
        },
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.6,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.6,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                    {
                      "detune": {
                        "inputs": [],
                        "value": 0,
                      },
                      "frequency": {
                        "inputs": [],
                        "value": 440,
                      },
                      "inputs": [],
                      "name": "OscillatorNode",
                      "type": "sine",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": -1,
          },
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which has an input node with no params 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 0.5,
          },
          "inputs": [
            {
              "detune": {
                "inputs": [],
                "value": 0,
              },
              "frequency": {
                "inputs": [],
                "value": 220,
              },
              "inputs": [],
              "name": "OscillatorNode",
              "type": "sine",
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes can define a custom node which has an input node with no params and using the string \`input\` instead of the \`INPUT\` identifier for backwards compatibility 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "gain": {
            "inputs": [],
            "value": 0.5,
          },
          "inputs": [
            {
              "detune": {
                "inputs": [],
                "value": 0,
              },
              "frequency": {
                "inputs": [],
                "value": 220,
              },
              "inputs": [],
              "name": "OscillatorNode",
              "type": "sine",
            },
          ],
          "name": "GainNode",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes creates a custom node which can be reused in virtualAudioGraph.update 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 0.5,
      },
      "inputs": [
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                            {
                              "detune": {
                                "inputs": [],
                                "value": 0,
                              },
                              "frequency": {
                                "inputs": [],
                                "value": 440,
                              },
                              "inputs": [],
                              "name": "OscillatorNode",
                              "type": "sine",
                            },
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": 1,
          },
        },
        {
          "inputs": [
            {
              "delayTime": {
                "inputs": [],
                "value": 0.5,
              },
              "inputs": [
                {
                  "gain": {
                    "inputs": [],
                    "value": 0.5,
                  },
                  "inputs": [
                    {
                      "delayTime": {
                        "inputs": [],
                        "value": 0.5,
                      },
                      "inputs": [
                        {
                          "gain": {
                            "inputs": [],
                            "value": 0.5,
                          },
                          "inputs": [
                            "<circular:DelayNode>",
                          ],
                          "name": "GainNode",
                        },
                      ],
                      "name": "DelayNode",
                    },
                    {
                      "detune": {
                        "inputs": [],
                        "value": 0,
                      },
                      "frequency": {
                        "inputs": [],
                        "value": 440,
                      },
                      "inputs": [],
                      "name": "OscillatorNode",
                      "type": "sine",
                    },
                  ],
                  "name": "GainNode",
                },
              ],
              "name": "DelayNode",
            },
          ],
          "name": "StereoPannerNode",
          "pan": {
            "inputs": [],
            "value": -1,
          },
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;

exports[`custom nodes updating startTime of nested custom nodes 1`] = `
{
  "inputs": [
    {
      "gain": {
        "inputs": [],
        "value": 1,
      },
      "inputs": [
        {
          "detune": {
            "inputs": [],
            "value": 0,
          },
          "frequency": {
            "inputs": [],
            "value": 440,
          },
          "inputs": [],
          "name": "OscillatorNode",
          "type": "sine",
        },
      ],
      "name": "GainNode",
    },
  ],
  "name": "AudioDestinationNode",
}
`;
